soroban-env-host/src/host/metered_clone.rs

```diff
@@ -121,82 +185,256 @@ impl MeteredClone for ScVal {
     }
 }
 
-impl MeteredClone for ScVec {
+impl MeteredClone for ScMapEntry {
     const IS_SHALLOW: bool = false;
+
+    const ELT_SIZE: u64 = 80;
+
     fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
-        for elt in self.0.iter() {
-            elt.charge_for_clone(budget)?
-        }
-        Ok(())
+        self.key.charge_for_clone(budget)?;
+        self.val.charge_for_clone(budget)
     }
 }
 
-impl MeteredClone for ScMap {
+impl MeteredClone for ScVec {
     const IS_SHALLOW: bool = false;
+
+    const ELT_SIZE: u64 = 24;
+
     fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
-        for elt in self.0.iter() {
-            elt.key.charge_for_clone(budget)?;
-            elt.val.charge_for_clone(budget)?;
-        }
-        Ok(())
+        self.0.charge_for_clone(budget) // self.0 will be deref'ed into Vec
     }
 }
 
-impl<const C: u32> MeteredClone for BytesM<C> {
+impl MeteredClone for ScMap {
     const IS_SHALLOW: bool = false;
 
+    const ELT_SIZE: u64 = 24;
+
     fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
-        charge_shallow_copy(1, budget)?;
-        budget.charge(CostType::BytesClone, self.len() as u64)
+        self.0.charge_for_clone(budget) // self.0 will be deref'ed into Vec
     }
 }
 
-impl MeteredClone for Vec<u8> {
+impl<const C: u32> MeteredClone for BytesM<C> {
     const IS_SHALLOW: bool = false;
 
+    const ELT_SIZE: u64 = 24;
+
     fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
-        charge_shallow_copy(1, budget)?;
-        budget.charge(CostType::BytesClone, self.len() as u64)
+        <Self as AsRef<Vec<u8>>>::as_ref(self).charge_for_clone(budget)
     }
 }
 
 impl<C: MeteredClone> MeteredClone for Vec<C> {
     const IS_SHALLOW: bool = false;
-    fn metered_clone(&self, budget: &Budget) -> Result<Self, HostError> {
-        if C::IS_SHALLOW {
-            // Shallow types we can batch-up our charging for.
-            budget.charge(CostType::BytesClone, self.len() as u64)?;
-            Ok(self.clone())
-        } else {
-            self.iter().map(|elt| elt.metered_clone(budget)).collect()
+
+    const ELT_SIZE: u64 = 24;
+
+    fn charge_for_clone(&self, budget: &Budget) -> Result<(), HostError> {
+        // first take care of Vec clone, which involves allocating memory, (shallowly) memcpy the
```

Comment:
Hm, this part I think isn't right -- it removed the special case of charging once for a clone of vectors of shallow elements.

Reply:
Except, wait, no, you _are_ charging for shallow copies of C, but then also charging for non-shallow cloning. I think I'm confused about what's going on here.